### ISSUE TITLE
Boundary mesh of submeshes

### ISSUE THREAD

**Issue by MG-82** [firedrakeproject/firedrake#2434]
This is basically a continuation of #723, #1282 and #1350. You even have a project for that, but its quite dead (is the submesh functionality really that exotic that nobody needs it?). I tried to fiddle around with the things in the branch that @ksagiyam created, but with the changes to PETSc in recent years, this is floating in mid air (the PETSc package by @wence- is still there). Kind of a pity, because the work already done seems to be quite elaborate and in an advanced stage.

I work with Claas (who opened #1282), and we encountered a problem where submesh functionality is no longer nice to have for us, but indeed critical. I can solve that partly with a nasty (and costly) workaround, but partly not (or I would have to implement an even nastier and costlier workaround).

Any chance that this gets reactivated? Thanks for consideration.

**Comment by ksagiyam** [comment]
Yes. We are currently wrapping up another project, but I actually started working on submesh again recently.

My plan is to first get the PETSc branch merged:
 
https://github.com/firedrakeproject/petsc/tree/dmplex_submesh_upstream

I updated/cleaned this branch a few month ago, but I still need to do some refactorisation.

I will then gradually merge the old Firedrake branch in to `master`.

Thanks!

**Comment by ksagiyam** [comment]
Just made a PETSc MR on submesh: https://gitlab.com/petsc/petsc/-/merge_requests/5626

Once this is merged (might take some time), I will start merging the old Firedrake submesh branch.
